### PR TITLE
Fix disconnection on some dongles!

### PR DESCRIPTION
**Pull request by freedreamer82**
I did some time ago  a PR adding connection Cb. I tested with some adapters but I noticed dbus worked in a different way...Connected property was not notified.
So I moved connection callback also in "add interface" method . I don't know what is related to this behavior  ..maybe different dongles and timing ?

another thing is the disconnection callback cannot have device as parameter 'casue I noticed can happen that dev is not ready anymore: so I just notified the user with a mac address string. If I think it sounds good because in connection cb you have a dev reference to work with but in the disconnection the device is gone.

Marco

### CODE REVIEW

File: bluezero/adapter.py [hunk 1]
```diff
@@ -288,6 +288,11 @@ def _interfaces_added(self, path, device_info):
                     adapter_addr=self.address,
                     device_addr=device_info[dev_iface]['Address'])
                 self.on_device_found(new_dev)
+            if device_info[dev_iface]['Connected'] and self.on_connect:
+                    new_dev = device.Device(
```
Comment (ukBaz) [inline bluezero/adapter.py]: This line is over indented

**Comment by ukBaz** [comment]
The InterfacesAdded signal and the PropertiesChanged signal do very different things.

Can you please give a scenario where the PropertiesChanged signal doesn't catch a connection event?

**Comment by freedreamer82** [comment]
when i connect to raspberry pi 3 for example the connection callback is not called and property changed neither.The strange thing is in InterfacesAdded passes the device with the Connected property.
never see something like that?

**Comment by freedreamer82** [comment]
I think i get it..when it's not paired the dongles works according to the current version. When I pair the phone on the connection dbus does not notify the property changed and only signals the interface adding.

let me know form your side.

marco

**Comment by ukBaz** [comment]
Hi Marco,

We are clearly doing something different because I'm seeing the "Connected" PropertiesChanged signal being sent both when pairing and connecting.

This is the minimal test I've been running:

```python
Python 3.7.3 (default, Oct  7 2019, 12:56:13) 
>>> from bluezero import adapter
>>> dongle = adapter.Adapter('FC:F8:AE:8F:0C:A4')
>>> dongle.on_connect = print
>>> dongle.run()
<bluezero.device.Device object at 0x7f61156b07f0>
<bluezero.device.Device object at 0x7f61153df588>
<bluezero.device.Device object at 0x7f61153dfda0>
<bluezero.device.Device object at 0x7f61153dfef0>
```
In another terminal I'm running `bluetoothctl` to do the pairing and connecting.

In yet another terminal I also had the following command running:
```
sudo busctl  monitor org.bluez | grep -A10 "Member=PropertiesChanged"
```
So you can see the raw signals also:
```
  Sender=:1.6  Path=/org/bluez/hci0/dev_DE_82_35_E7_43_BE  Interface=org.freedesktop.DBus.Properties  Member=PropertiesChanged
  UniqueName=:1.6
  MESSAGE "sa{sv}as" {
          STRING "org.bluez.Device1";
          ARRAY "{sv}" {
                  DICT_ENTRY "sv" {
                          STRING "UUIDs";
                          VARIANT "as" {
                                  ARRAY "s" {
                                          STRING "00001800-0000-1000-8000-00805f9b34fb";
                                          STRING "00001801-0000-1000-8000-00805f9b34fb";
```

**Comment by freedreamer82** [comment]
Hi,
I followed your steps on a raspberry pi 3.The test uses the lib in peripheral mode.
when i connect to raspberry with my phone i got this:

```
  File "/usr/lib/python2.7/dist-packages/dbus/connection.py", line 230, in maybe_handle_message
    self._handler(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/bluezero/adapter.py", line 267, in _properties_changed
    device_addr=macaddr)
  File "/usr/local/lib/python2.7/dist-packages/bluezero/device.py", line 56, in __init__
    " using adapter: " + adapter_addr)
ValueError: Cannot find a device: 54:60:8E:18:4A:8C using adapter: B8:27:EB:2F:3C:FB

```
and monitoring dbus seems no change in property

>   Sender=:1.5  Path=/org/bluez/hci0/dev_54_60_8E_18_4A_8C  Interface=org.freedesktop.DBus.Properties  Member=PropertiesChanged
  UniqueName=:1.5
  MESSAGE "sa{sv}as" {
          STRING "org.bluez.Device1";
          ARRAY "{sv}" {
                  DICT_ENTRY "sv" {
                          STRING "ServicesResolved";
                          VARIANT "b" {
                                  BOOLEAN false;
                          };
                  };

>  Sender=:1.5  Path=/org/bluez/hci0/dev_54_60_8E_18_4A_8C  Interface=org.freedesktop.DBus.Properties  Member=PropertiesChanged
  UniqueName=:1.5
  MESSAGE "sa{sv}as" {
          STRING "org.bluez.Device1";
          ARRAY "{sv}" {
                  DICT_ENTRY "sv" {
                          STRING "UUIDs";
                          VARIANT "as" {
                                  ARRAY "s" {
                                          STRING "00001800-0000-1000-8000-00805f9b34fb";
                                          STRING "00001801-0000-1000-8000-00805f9b34fb";

>   Sender=:1.5  Path=/org/bluez/hci0/dev_54_60_8E_18_4A_8C  Interface=org.freedesktop.DBus.Properties  Member=PropertiesChanged
  UniqueName=:1.5
  MESSAGE "sa{sv}as" {
          STRING "org.bluez.Device1";
          ARRAY "{sv}" {
                  DICT_ENTRY "sv" {
                          STRING "Name";
                          VARIANT "s" {
                                  STRING "iPhone";
                          };
                  };

>  Sender=:1.5  Path=/org/bluez/hci0/dev_54_60_8E_18_4A_8C  Interface=org.freedesktop.DBus.Properties  Member=PropertiesChanged
  UniqueName=:1.5
  MESSAGE "sa{sv}as" {
          STRING "org.bluez.Device1";
          ARRAY "{sv}" {
                  DICT_ENTRY "sv" {
                          STRING "Appearance";
                          VARIANT "q" {
                                  UINT16 64;
                          };
                  };

**Comment by ukBaz** [comment]
OK, I had missed that you were using the RPi in a peripheral role.

My test was for the central role. Do you have some code you can share? Or a simple testcase that you are using?

**Comment by ukBaz** [comment]
Hi @freedreamer82 ,

I have been able to reproduce this when I am running a peripheral so I can now test this. Don't have time to look at it today but I am still working on this.

Thanks,
Barry

**Comment by freedreamer82** [comment]
thanks! Let me know..

MArco

**Comment by freedreamer82** [comment]
hi, news?

**Comment by ukBaz** [comment]
Hi,

Apologies for no updates. Life is rather hectic at the moment and I this cannot priority right now. I have however been making some progress, just no real conclusions.

This issue does seem to be related to the difference between a peripheral or central roles.
A peripheral can only be connected to one central device (such as a mobile phone) at a time, but the central device can be connected to multiple peripherals.
So when you are in peripheral mode and accept a connection the device seems to get created and destroyed on connection/disconnect. While when you are in central mode and you connect to a known device, that device is persistent in the Bluez database so Bluez changes the connection status.
I've been trying to find some Bluez documentation that corroborates this.

My thought is that `on_connected` shouldn't be used when in peripheral mode and it should be something else. However, I don't know what that term should be.

Thoughts?

**Comment by ukBaz** [comment]
This has been sat here for a couple of years without the issue getting to a full conclusion and being merged. The code has now moved on and there are merge conflicts. As a result I'm going to go ahead and close this.
Hopefully someone will come up with a fix for this. And maybe more importantly, a good way to test this between central and peripheral.

**Pull request closed without merging**